gk/main.c

```diff
@@ -546,44 +613,92 @@ gk_proc(void *arg)
 
 			/* 
 			 * Find the flow entry for the IP pair.
-			 * Create a new flow entry if not found.
+			 *
+			 * If the pair of source and destination addresses 
+			 * is in the flow table, proceed as the entry instructs,
+			 * and go to the next packet.
 			 */
 			ret = rte_hash_lookup_with_hash(
 				instance->ip_flow_hash_table,
 				&packet.flow, pkt->hash.rss);
-			if (ret < 0) {
-				/* Create a new flow entry. */
-				ret = rte_hash_add_key_with_hash(
-					instance->ip_flow_hash_table,
- 					(void *)&packet.flow, pkt->hash.rss);
-				if (ret < 0) {
-					RTE_LOG(ERR, HASH,
-						"The GK block failed to add new key to hash table!\n");
-					rte_pktmbuf_free(pkt);
+			if (ret >= 0)
+				fe = &instance->ip_flow_entry_table[ret];
+			else {
+				/*
+				 * Otherwise, look up the destination address
+			 	 * in the global LPM table.
+				 */
+				struct gk_policy *policy = get_gk_policy(
+					&gk_conf->lpm_tbl, &packet.flow);
+
+			 	/*
+				 * No entry for the destination,
+				 * drop the packet.
+				 */
+				if (policy == NULL) {
+					drop_packet(pkt);
 					continue;
 				}
 
-				fe = &instance->ip_flow_entry_table[ret];
-				initialize_flow_entry(fe, &packet.flow);
-			} else
-				fe = &instance->ip_flow_entry_table[ret];
+				switch (policy->action) {
+				case GK_FWD_GT:
+					/*
+			 		 * The entry instructs to enforce
+					 * policies over its packets,
+			 		 * initialize an entry in the
+					 * flow table, proceed as the
+					 * brand-new entry instructs, and
+			 		 * go to the next packet.
+			 		 */
+					ret = rte_hash_add_key_with_hash(
+						instance->ip_flow_hash_table,
+ 						(void *)&packet.flow,
+						pkt->hash.rss);
+					if (ret < 0) {
+						RTE_LOG(ERR, HASH,
+						"The GK block failed to add new key to hash table!\n");
+						drop_packet(pkt);
+						continue;
+					}
+
+					fe = &instance->
+						ip_flow_entry_table[ret];
+					fe->u.request.grantor_id = get_grantor_id(
+						&gk_conf->lpm_tbl, &packet.flow);
+					initialize_flow_entry(fe, &packet.flow, &gk_conf->lpm_tbl);
```

Comment:
`initialize_flow_entry()` is going to find the `grantor_id`, so the lookup just before the call to `initialize_flow_entry()` is useless.